Add blog posts about WebAssembly in Cloud Native

cr7258 opened the pull request:
This blog post walks through building and deploying Wasm applications in the Cloud Native ecosystem.

cr7258 commented on content/en/blog/_posts/2023-11-11-webassembly-in-cloud-native/index.md:
````diff
@@ -0,0 +1,857 @@
+docker build -t docker.io/cr7258/wasm-demo-app:v1 .
+```
+
+To make it easier to use in the following tutorials, push the image to Docker Hub. Replace the repo  with your own.
````
I change `tutorials` to `demos`: https://github.com/kubernetes/website/pull/43760/commits/4a8566e32ad9e257a74b210f2c0f36be1e89f274

cr7258 commented on content/en/blog/_posts/2023-11-11-webassembly-in-cloud-native/index.md:
````diff
@@ -0,0 +1,857 @@
+Install WasmEdge.
+
+```bash
+curl -sSf https://raw.githubusercontent.com/WasmEdge/WasmEdge/master/utils/install.sh | bash
````
Add comment to remind about the security implications: https://github.com/kubernetes/website/commit/4a8566e32ad9e257a74b210f2c0f36be1e89f274

cr7258 commented on content/en/blog/_posts/2023-11-11-webassembly-in-cloud-native/index.md:
```diff
@@ -0,0 +1,857 @@
+
+### Install Rust
+
+ Install Rust by running the following command in your terminal.
```
hyperlink to Rust's page: https://github.com/kubernetes/website/commit/4a8566e32ad9e257a74b210f2c0f36be1e89f274

cr7258 commented on content/en/blog/_posts/2023-11-11-webassembly-in-cloud-native/index.md:
````diff
@@ -0,0 +1,857 @@
+```
+##### KWasm Operator
+
+In this section, I will demostrate how to use Kwasm Operator to add Wasm support to Kubernetes nodes automatically. To enable Wasm support on a particular node, simply add the annotation `kwasm.sh/kwasm-node=true` on that node. This will trigger Kwasm to create a Job to deploy the necessary binary files needed to run Wasm on the node. Additionally, containerd's configuration will be modified acco ...
````
Add description for Kwasm Operator: https://github.com/kubernetes/website/pull/43760/commits/09ddb7a64bc33b87757c247afb8e69e61b4d6627

cr7258 commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
```diff
@@ -0,0 +1,858 @@
+
+In this section, I will demonstrate the manual installation of crun with the WasmEdge runtime library, and adjust containerd config to use crun as the runtime, enabling Wasm support on the Kubernetes node.
+
+Create a single-node Kubernetes cluster using kind.
```
I mentioned here there is only 1 node in the cluster. So we don't need to add node selector here?

sftim replied:
You can instead explain that this advice only makes sense in this specific context and won't be generally applicable. Pick whichever you like of those 2 options (or suggest another). The aim is to make sure that readers don't infer something unhelpful.

cr7258 replied:
Ok. I add nodeSelector for the pod: https://github.com/kubernetes/website/pull/43760/commits/4d7edac253c031fbff397678d467cbd64d7e562d

cr7258 commented on content/en/blog/_posts/2023-11-11-webassembly-in-cloud-native/index.md:
```diff
@@ -0,0 +1,857 @@
+  runtimeClassName: crun
+  containers:
+  - name: wasm-demo-app
+    image: docker.io/cr7258/wasm-demo-app:v1
```
reply: https://github.com/kubernetes/website/pull/43760#discussion_r1378477945

cr7258 commented:
> Also, please add alt attributes for each figure.

Add alt attributes: https://github.com/kubernetes/website/pull/43760/commits/a8cc4f86f48258e7104733cc60e5f5c88beec3e7

> please use SVG diagrams, ideally following our [guide](https://kubernetes.io/docs/contribute/style/diagram-guide/)

I tried to use SVG, but I found the font in the diagram is modified. Could I still use PNG files?

svg:
<img width="1005" alt="image" src="https://github.com/kubernetes/website/assets/40051120/436b83f1-648b-490f-bc34-58e79720cb03">

png:
<img width="998" alt="image" src="https://github.com/kubernetes/website/assets/40051120/90df868b-4e08-4005-a361-d9ad1e6f6902">

sftim commented:
What's the source format for those images? Maybe we can help you to export scalable versions.

sftim commented:
For the font, we recommend converting text to curves. We can also give you advice on how to do that.

cr7258 commented:
I drew these images on https://excalidraw.com/. I uploaded the file, and you can download it through this link: https://www.transfernow.net/en/d?utm_source=20231101SCofBSGq. 

you can select the image and right-click, then copy the image as SVG.

<img width="1737" alt="image" src="https://github.com/kubernetes/website/assets/40051120/7048eed7-91c7-4691-8b73-fde643b1c42d">

Could you please help me take a look at it? Thanks a lot.

leonardpahlke commented:
Hey, all. Shimmed in to support and get this blog moved forward 🙌 . I exported the images to `.svg` format. I used the standard font family from excalidraw. 

!
![01-webassembly-runs-on-browser-and-server](https://github.com/kubernetes/website/assets/58054077/cacd8e7d-bf01-4277-8102-2de5362a7133)
[02-high-level-and-low-level-container-runtimes](https://github.com/kubernetes/website/assets/58054077/32f2d5c3-0f88-4ede-b1f1-0127de3b8a4d)
![05-runwasi-is-the-bridge-between-containerd-and-wasm-runtime](https://github.com/kubernetes/website/assets/58054077/83224241-d24f-46c8-833a-875d7de3318d)
![04-use-containerd-shim-to-manage-wasm-modules](https://github.com/kubernetes/website/assets/58054077/57feb5b0-3c44-4b5c-913c-5a3fc6faf2c9)
![07-kwasm-operator](https://github.com/kubernetes/website/assets/58054077/a6604244-3119-4277-bdb9-f765793a9ba7)


Converted screenshot from `.png` to `.svg` (using Adobe Express):
![06-enable-wasm-support-on-docker-desktop](https://github.com/kubernetes/website/assets/58054077/35c63940-c024-42f5-9e32-fb1bf752de3e)

@cr7258 since the quality of the screenshot in svg format is not that good, I would consider cutting the screenshot for a short one sentence explanation with a link to the docker + wasm [documentation](https://docs.docker.com/desktop/wasm/). To me the screenshot is not essential to the blog, providing minimal info.

cr7258 commented:
@leonardpahlke Many thanks! 🙌🙌🙌
I replaced the PNG file with the SVG you provided.

> since the quality of the screenshot in svg format is not that good, I would consider cutting the screenshot for a short one sentence explanation with a link to the docker + wasm [documentation](https://docs.docker.com/desktop/wasm/). To me the screenshot is not essential to the blog, providing minimal info.

I removed the screenshot and used a short explanation with a link to the documentation.

leonardpahlke commented:
Great! As far as I can see, all comments are closed. Should be ready to go.

@sftim PTAL.

sftim commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
````diff
@@ -0,0 +1,864 @@
+Hello, World!
+```
+
+## Run Wasm Modules in the Container Ecosystem
+
+### Run Wasm Modules in Linux Container
````
```suggestion
## Run Wasm modules using container tooling

### Run Wasm modules in a local Linux container, manually
```

sftim commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
````diff
@@ -0,0 +1,864 @@
+CMD ["wasmedge", "--dir", ".:/", "/http-server.wasm"]
+```
+
+Build the container image.
````
```suggestion
Build the container image:
```

sftim commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
```diff
@@ -0,0 +1,864 @@
+
+Next, I will individually demonstrate how to run Wasm modules through both low-level and high-level container runtimes.
+
+#### Low-level Container Runtime
```
How about:
```suggestion
#### Run Wasm modules in a local Linux container, via a low-level runtime
```

sftim commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
````diff
@@ -0,0 +1,864 @@
+     go-md2man libtool autoconf python3 automake
+```
+
+Configure, build, and install a crun binary with WasmEdge support.
````
```suggestion
Configure, build, and install a `crun` binary that includes WasmEdge support:
```
?

sftim commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
````diff
@@ -0,0 +1,864 @@
+}
+```
+
+Start the container using crun. 
````
```suggestion
Start the container using `crun`:
```

sftim commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
```diff
@@ -0,0 +1,864 @@
+The _container shim_ serves as a bridge between high and low-level container runtimes. Its main purpose is to abstract low-level runtime details, enabling uniform management of various runtimes in high-level runtime. In this section, I will use _[containerd](https://github.com/containerd/containerd)_ as an example. Containerd is an industry-standard container runtime with an emphasis on simplicit ...
+
+Containerd can manage Wasm modules in two ways:
+- 1.Manages Wasm modules through container runtimes like crun and youki that support building with the Wasm runtime library. These two runtimes can also run regular Linux containers. Containerd uses `containerd-shim-runc-v2` to interface with low-level container runtimes.
+- 2.Manages Wasm modules directly through Wasm runtimes, such as WasmEdge, Wasmtime, Spin, and Slight. Containerd uses `containerd-wasm-shim(runwasi)` to interface with Wasm runtimes.
```
We list products from the ecosystem in either (fair) random order, or alphabetical order.
```suggestion
1. Manages Wasm modules through container runtimes like crun and youki that support building with the Wasm runtime library. These two runtimes can also run regular Linux containers. Containerd uses `containerd-shim-runc-v2` to interface with low-level container runtimes.
1. Manages Wasm modules directly through Wasm runtimes, such as Slight, Spin, WasmEdge, and Wasmtime. Containerd uses `containerd-wasm-shim(runwasi)` to interface with Wasm runtimes.
```

BTW, Hugo can do the counting for us hence both items are numbered as 1.

sftim commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
```diff
@@ -0,0 +1,864 @@
+In this section, I will demonstrate how to configure crun as runtime in containerd, enabling support for running Wasm modules.
+
+{{< note >}}
+Ensure that crun binary has been installed as per the instructions in the section: [Low-level Container Runtime](#low-level-container-runtime).
```
You might need to update the link here if following feedback elsewhere.

cr7258 replied:
Updated the link: https://github.com/kubernetes/website/pull/43760/commits/d4ee5b15713b53f4baa41c365d2a36d6533f4aa3

sftim commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
```diff
@@ -0,0 +1,864 @@
+Ensure that crun binary has been installed as per the instructions in the section: [Low-level Container Runtime](#low-level-container-runtime).
+{{< /note >}}
+
+Run the following commands to install containerd.
```
```suggestion
Run the following commands to install containerd:
```

sftim commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
```diff
@@ -0,0 +1,864 @@
+wget https://github.com/containerd/containerd/releases/download/v${VERSION}/cri-containerd-cni-${VERSION}-linux-amd64.tar.gz.sha256sum
+sha256sum --check cri-containerd-cni-${VERSION}-linux-amd64.tar.gz.sha256sum
+
+sudo tar --no-overwrite-dir -C / -xzf cri-containerd-cni-${VERSION}-linux-amd64.tar.gz
```
This looks dangerous; why deploy to `/`?

cr7258 replied:
The files extracted from the archive are as follows, including containerd binary, cni plugins, config file, etc. it will place the files from `/`, for example: `/usr/local/bin/containerd`.
```
├── cri-containerd.DEPRECATED.txt
├── etc
│   ├── cni
│   │   └── net.d
│   │       └── 10-containerd-net.conflist
│   ├── crictl.yaml
│   └── systemd
│       └── system
│           └── containerd.service
├── opt
│   ├── cni
│   │   └── bin
│   │       ├── bandwidth
│   │       ├── bridge
│   │       ├── dhcp
│   │       ├── dummy
│   │       ├── firewall
│   │       ├── host-device
│   │       ├── host-local
│   │       ├── ipvlan
│   │       ├── loopback
│   │       ├── macvlan
│   │       ├── portmap
│   │       ├── ptp
│   │       ├── sbr
│   │       ├── static
│   │       ├── tuning
│   │       ├── vlan
│   │       └── vrf
│   └── containerd
│       └── cluster
│           ├── gce
│           │   ├── cloud-init
│           │   │   ├── master.yaml
│           │   │   └── node.yaml
│           │   ├── cni.template
│           │   ├── configure.sh
│           │   └── env
│           └── version
└── usr
    └── local
        ├── bin
        │   ├── containerd
        │   ├── containerd-shim
        │   ├── containerd-shim-runc-v1
        │   ├── containerd-shim-runc-v2
        │   ├── containerd-stress
        │   ├── crictl
        │   ├── critest
        │   ├── ctd-decoder
        │   └── ctr
        └── sbin
            └── runc

16 directories, 38 files
```

sftim commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
```diff
@@ -0,0 +1,864 @@
+
+wget https://github.com/containerd/containerd/releases/download/v${VERSION}/cri-containerd-cni-${VERSION}-linux-amd64.tar.gz
+wget https://github.com/containerd/containerd/releases/download/v${VERSION}/cri-containerd-cni-${VERSION}-linux-amd64.tar.gz.sha256sum
+sha256sum --check cri-containerd-cni-${VERSION}-linux-amd64.tar.gz.sha256sum
```
Extra credit: list the checksum you expect.

cr7258 replied:
Added expected checksum: https://github.com/kubernetes/website/pull/43760/commits/5cc1917d7948ee37904f81f03a4b9df713682c9b

sftim commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
```diff
@@ -0,0 +1,864 @@
+
+sudo tar --no-overwrite-dir -C / -xzf cri-containerd-cni-${VERSION}-linux-amd64.tar.gz
+sudo systemctl daemon-reload
+sudo systemctl start containerd
```
(nit)
I'd write
```suggestion
sudo systemctl enable containerd --now
```

sftim commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
```diff
@@ -0,0 +1,864 @@
+
+You can run Wasm modules through containerd:
+- `--runc-binary`: Specifies to use `crun` as low-level runtime to start the container.
+- `--runtime`: Specifies the version and name of the shim, which are converted by containerd into the binary name of the shim: `io.containerd.runc.v2 -> containerd-shim-runc-v2`. Containerd starts the shim by running `containerd-shim-runc-v2` binary file, which subsequently invokes crun to launch the container.
```
(nits)
```suggestion
- `--runtime`: Specifies the version and name of the shim, which are converted by containerd into the binary name of the shim. For example: `io.containerd.runc.v2` → `containerd-shim-runc-v2`. Containerd starts the shim by running the `containerd-shim-runc-v2` binary file, which subsequently invokes `crun` to launch the container.
```

sftim commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
````diff
@@ -0,0 +1,864 @@
+kubectl port-forward pod/wasm-demo-app 8080:8080
+```
+
+Open a new terminal, send a request to the service.
````
```suggestion
Send a request to the service running in the test Pod inside Kubernetes:
```

cr7258 replied:
I have made the service available locally using `kubectl port-forward`, and I can access it through the URL http://localhost:8080. So there is no need to send the request in the test Pod inside Kubernetes?

sftim commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
````diff
@@ -0,0 +1,864 @@
+    runtime: crun
+handler: crun
+```
+
````
```suggestion

##### Run the WebAssembly app via Kubernetes

Next, run the Wasm app inside a Kubernetes pod.
```
?

cr7258 replied:
I don't add this new header. I changed the header and divided running wasm on kubernetes into manual and automatic using Kwasm.

<img width="340" alt="image" src="https://github.com/kubernetes/website/assets/40051120/7296bfe0-de61-42ca-9732-f834a73e74d5">

sftim commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
````diff
@@ -0,0 +1,864 @@
+```
+
+Create a RuntimeClass resource named `crun` to use the crun handler automatically set up by Kwasm in containerd, the `scheduling.nodeSelector` property sends pod to nodes with the `runtime=crun` label.
+
````
```suggestion
Here is an example manifest:

```

sftim commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
````diff
@@ -0,0 +1,864 @@
+```bash
+kind delete cluster --name wasm-demo
+```
+##### KWasm Operator
````
This could be:
```markdown
##### Set up the KWasm Operator
```
or
```suggestion
##### Set up your cluster for Wasm using KWasm, then run the app inside a pod
```

sftim commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
````diff
@@ -0,0 +1,864 @@
+EOF
+```
+
+Restart containerd.
````
```suggestion
Restart containerd:
```

sftim commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
````diff
@@ -0,0 +1,864 @@
+systemctl restart containerd
+```
+
+Add label `runtime=crun` on the node.
````
```suggestion
Set the label `runtime=crun` on the node:
```

sftim commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
````diff
@@ -0,0 +1,864 @@
+    image: docker.io/cr7258/wasm-demo-app:v1
+```
+
+You can use `port-forward` to forward traffics to your local machine for access.
````
I'd write:
```suggestion
You can use `kubectl port-forward` to forward traffic from your local machine into the Kubernetes cluster:
```

sftim commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
````diff
@@ -0,0 +1,864 @@
+kind delete cluster --name kwasm-demo
+```
+
+In this article, the `module.wasm.image/variant: compat` annotation is used to indicate to the container runtime that the workload is a Wasm workload. In this [PR](https://github.com/containers/crun/pull/886), crun has introduced a new annotation: `module.wasm.image/variant: compat-smart`. 
````
Although https://github.com/containers/crun/pull/886 calls it an annotation, I think the technically correct name is “label”. When it's set on a container it's a label.

sftim commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
````diff
@@ -0,0 +1,864 @@
+kubectl port-forward pod/wasm-demo-app 8080:8080
+```
+
+Open a new terminal, send a request to the service.
````
```suggestion
Open a new terminal, and send a request to the service (running in the test Pod inside Kubernetes):
```

cr7258 replied:
I have made the service available locally using `kubectl port-forward`, and I can access it through the URL http://localhost:8080. So there is no need to send the request in the test Pod inside Kubernetes?

cr7258 commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
```diff
@@ -0,0 +1,864 @@
+### Run Wasm modules on Container Management Platforms
+
+#### Run Wasm modules on Docker Desktop
+
```
Added: https://github.com/kubernetes/website/pull/43760/commits/d4ee5b15713b53f4baa41c365d2a36d6533f4aa3

cr7258 commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
```diff
@@ -0,0 +1,864 @@
+
+WebAssembly is [driving the third wave of cloud computing](https://nigelpoulton.com/webassembly-the-future-of-cloud-computing/). As the de facto standard in the realm of container orchestration, Kubernetes continuously evolves to leverage the advantages brought about by WebAssembly.
+
+The major way to run Wasm modules on Kubernetes is adding Wasm support to the Kubernetes nodes. To achieve this, we can either manually install a container runtime build with the Wasm runtime library or use the Kwasm Operator to automate the process.
```
Use `you` and `built`: https://github.com/kubernetes/website/pull/43760/commits/d4ee5b15713b53f4baa41c365d2a36d6533f4aa3

cr7258 commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
````diff
@@ -0,0 +1,864 @@
+
+Seeing `+WASM:wasmedge` indicates that WasmEdge has been installed in crun.
+
+```bash
````
Remove the syntax marker: https://github.com/kubernetes/website/pull/43760/commits/d4ee5b15713b53f4baa41c365d2a36d6533f4aa3

cr7258 commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
````diff
@@ -0,0 +1,864 @@
+
+Edit the `Cargo.toml` file to add the dependencies listed below. _[warp_wasi](https://crates.io/crates/warp_wasi)_ is specifically designed for WASI and is built upon the _[Warp](https://docs.rs/warp/latest/warp/)_ framework, which is a lightweight web server framework used to develop high-performance asynchronous web applications.
+
+```bash
````
Remove the syntax marker: https://github.com/kubernetes/website/pull/43760/commits/d4ee5b15713b53f4baa41c365d2a36d6533f4aa3

cr7258 commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
````diff
@@ -0,0 +1,864 @@
+
+```bash
+kubectl label nodes kwasm-demo-control-plane runtime=crun
+```
````
Use `crun-kwasm`: https://github.com/kubernetes/website/pull/43760/commits/b2ecd7fbe57fd9d5398244c899faeb4f69ace7fd

cr7258 commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
````diff
@@ -0,0 +1,864 @@
+```
+
+##### Kubernetes + Containerd + Crun
+
````
Here is a note to ask the user to follow the instructions in the previous section to install crun with Wasm support: https://github.com/kubernetes/website/blob/b2ecd7fbe57fd9d5398244c899faeb4f69ace7fd/content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md?plain=1#L692

cr7258 commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
```diff
@@ -0,0 +1,864 @@
+
+In this article, I use _[WasmEdge](https://wasmedge.org/docs/)_, a lightweight, high-performance, and extensible WebAssembly runtime. 
+
+Install WasmEdge.
```
Change it to `Install WasmEdge by running:`. 
The header `### Install WasmEdge` already exists: https://github.com/kubernetes/website/blob/b2ecd7fbe57fd9d5398244c899faeb4f69ace7fd/content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md?plain=1#L101-L107

cr7258 commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
````diff
@@ -0,0 +1,864 @@
+kind create cluster --name wasm-demo
+```
+
+Each Kubernetes node created by kind is a Docker container, you can enter that node using the `docker exec` command.
````
Updated: https://github.com/kubernetes/website/pull/43760/commits/1c7e53b2bd14cde40ce7b89e0b8725be9cd27ff8

cr7258 commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
```diff
@@ -0,0 +1,864 @@
+
+**Author:** Seven Cheng (SAP)
+
+_[WebAssembly](https://webassembly.org/)_ (Wasm) was originally created for the browser, and it has become increasingly popular on the server-side as well. WebAssembly is gaining popularity in the Cloud Native ecosystem due to its advantages over containers, including smaller size, faster speed, enhanced security, and greater portability. 
```
Updated: https://github.com/kubernetes/website/commit/1c7e53b2bd14cde40ce7b89e0b8725be9cd27ff8

cr7258 commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
```diff
@@ -0,0 +1,864 @@
+metadata:
+  name: wasm-demo-app
+  annotations:
+    module.wasm.image/variant: compat-smart
```
https://github.com/kubernetes/website/pull/43760/commits/a1174d88560edf4b5dc379fd22d8824d4f3d5320

0xE282B0 commented on content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md:
````diff
@@ -0,0 +1,880 @@
+# Provision Nodes
+kubectl annotate node kwasm-demo-control-plane kwasm.sh/kwasm-node=true
+```
+
+Add label `runtime=crun-kwasm` on the node.
+
+```bash
+kubectl label nodes kwasm-demo-control-plane runtime=crun-kwasm
+```
+
+Create a RuntimeClass resource named `crun` to use the crun handler automatically set up by Kwasm in containerd, the `scheduling.nodeSelector` property sends pod to nodes with the `runtime=crun-kwasm` label. Here is an example manifest:
+
+```yaml
+apiVersion: node.k8s.io/v1
+kind: RuntimeClass
+metadata:
+  name: crun
+scheduling:
+  nodeSelector:
+    runtime: crun-kwasm
+handler: crun
+```
````
@cr7258, @sftim, IMPORTANT NOTE:
KWasm by default uses the latest [https://github.com/KWasm/kwasm-node-installer](kwasm-node-installer) image built from the `main' branch. Which at the time of writing is > 0.3.0 

> [!CAUTION]
> KWasm Node Installer version v0.3.0 has removed crun in favor of the WasmEdge shim.

There are now two options:

1. Use the v0.2.0 version of the installer.
2. Change the RuntimeClass to `wasmedge`.

## Using the v0.2.0 installer image
```
helm upgrade --install -n kwasm --create-namespace kwasm-operator kwasm/kwasm-operator \
  --set kwasmOperator.autoProvision="true" \
  --set kwasmOperator.installerImage="ghcr.io/kwasm/kwasm-node-installer:v0.2.0"
```

## Changing the RuntimeClass
The `wasmedge` shim behaves like `crun+wasmedge`, but it doesn't need the `compat/compat-smart` annotation. It always uses the Linux container as a fallback. All images that work with `crun+wasmedge` should also work with the shim.

```yaml
apiVersion: node.k8s.io/v1
kind: RuntimeClass
metadata:
  name: wasmedge
handler: wasmedge
```

## Reasoning
Maintaining the installation of different runtimes on different Linux and Kubernetes distributions is a complex task. KWasm will be redesigned to better address the requirement to install as many runtimes as possible, but at this stage only runwasi-based shims will be shipped because they have the simplest installation, while most wasm runtimes are available as runwasi shims.

cr7258 replied:
@0xE282B0 Thanks for your reminder. I update content related to KWasm. https://github.com/kubernetes/website/pull/43760/commits/a2cbd9575a0dd6738ae400bc19becc0a63085146

0xE282B0 reviewed (commented):
Great article with a nice journey along the different ways to run Wasm images.
Unfortunately the KWasm demo is outdated, crun is not shipped anymore after node-installer v0.3.0.
I left a detailed comment on that. Let me know if I can help with something 😊

sftim commented:
/hold
The blog team need to assign a publication date and have this PR updated (2 places) to match that assigned date.

OK to unhold once that has happened.

divya-mohan0209 commented:
My two cents on this blog piece: There's a lot of information being conveyed in one post & I'd recommend that we split it into two sections.

My recommendation : First blog post deals with the theory of Wasm & Wasm in Cloud Native (& the other would deal with the practical implementation of the concepts. I'm open to hearing if @cr7258 or anyone else has alternate suggestions.

Additionally, I'd recommend we push this to January for publication because readership tends to be on the lower side during December and it'd also be a challenge from a bandwidth perspective.

nate-double-u commented:
> Additionally, I'd recommend we push this to January for publication because readership tends to be on the lower side during December and it'd also be a challenge from a bandwidth perspective.

+1, pushing till January will also give this post (or posts if you decide to split it) some distance from the release posts that are planned for mid Dec. I'd suggest 2nd week of January, maybe the  8-10.

cr7258 commented:
Hi, @divya-mohan0209  @nate-double-u. Thanks for your recommendation. 
>  First blog post deals with the theory of Wasm & Wasm in Cloud Native (& the other would deal with the practical implementation of the concepts. 

I feel that this article is basically a practical guide. For the sake of completeness, only the first two chapters provide a brief introduction to the background of Wasm. So, I prefer not to split this article.
<img width="1563" alt="image" src="https://github.com/kubernetes/website/assets/40051120/353c70ce-374d-4b5f-8f48-d86bdfebf547">

> +1, pushing till January will also give this post (or posts if you decide to split it) some distance from the release posts that are planned for mid Dec. I'd suggest 2nd week of January, maybe the 8-10.

I'm fine with the date of publication.

nate-double-u commented:
> I feel that this article is basically a practical guide. For the sake of completeness, only the first two chapters provide a brief introduction to the background of Wasm. So, I prefer not to split this article.

But it can be expanded—just cutting the article in half wasn't @divya-mohan0209's suggestion

cr7258 commented:
Got it. Thanks for all of your suggestions. 

I will split the article into two sections. I plan to finish it in about 1 week.
> First blog post deals with the theory of Wasm & Wasm in Cloud Native (& the other would deal with the practical implementation of the concepts.

sftim commented:
@cr7258, what help could we offer you to move this article series forward?

cr7258 commented:
Hi @sftim, I split the article into two sections. Please review them, Thanks. 
I set the publication date at the end of January (January 26). Sorry for the delay.

cr7258 commented:
Hi @sftim 
Can you please have a look at this article?

Gauravpadam commented on content/en/blog/_posts/2024-01-26-webassembly-on-kubernetes-from-containers-to-wasm/index.md:
```diff
@@ -0,0 +1,73 @@
+
+Generally, container runtimes can be categorized into two levels: high-level runtimes and low-level runtimes.
+- **Low-level Container Runtime**: This refers to OCI-compliant implementations that can receive a runnable filesystem (rootfs) and a configuration file (config.json) to execute isolated processes. Low-level container runtimes directly manage and run containers, such as runc, crun, youki, gvisor, and kata.
+- **High-level Container Runtime**: This is responsible for the transport and management of container images, unpacking the image, and passing it off to the low-level runtime to run the container. High-level container runtimes simplify container management by abstracting the complexities of low-level runtime, which allows users to manage various low-level runtimes via the same high-level runtime. ...
```
```suggestion
- **High-level Container Runtime**: This is responsible for the transport and management of container images, unpacking the image, and passing it off to the low-level runtime to run the container. High-level container runtimes simplify container management by abstracting the complexities of low-level runtime, which allows users to manage various low-level runtimes through the same high-level runtime. Containerd and CRI-O are two popular high-level container runtimes.
```

Gauravpadam reviewed (commented):
Thanks

A few nits

Gauravpadam commented on content/en/blog/_posts/2024-01-26-webassembly-on-kubernetes-the-practice-guide/index.md:
```diff
@@ -0,0 +1,845 @@
+
+### Run Wasm modules via low-level container runtimes
+
+_[Crun](https://github.com/containers/crun)_ is a fast and lightweight OCI container runtime written in C, which has built-in support for WasmEdge. In this section, I will demonstrate how to utilize crun to directly launch a Wasm module using the provided config.json and rootfs files, without depending on high-level container runtimes.
```
```suggestion
_[Crun](https://github.com/containers/crun)_ is a fast and lightweight OCI container runtime written in C, which has built-in support for WasmEdge. In this section, I will demonstrate how to utilize crun to directly launch a Wasm module using the provided `config.json` and `rootfs` files, without depending on high-level container runtimes.
```

I suggest refactoring a few places in the next commit

Gauravpadam commented on content/en/blog/_posts/2024-01-26-webassembly-on-kubernetes-the-practice-guide/index.md:
```diff
@@ -0,0 +1,845 @@
+The _container shim_ serves as a bridge between high-level and low-level container runtimes. Its main purpose is to abstract low-level runtime details, enabling uniform management of various low-level runtimes in high-level runtime. In this section, I will use _[containerd](https://github.com/containerd/containerd)_ as an example. Containerd is an industry-standard container runtime with an empha ...
+
+Containerd can manage Wasm modules in two ways:
+1. Manages Wasm modules through container runtimes like crun and youki that support building with the Wasm runtime library. These two runtimes can also run regular Linux containers. Containerd uses `containerd-shim-runc-v2` to interface with low-level container runtimes.
```
```suggestion
1. Through container runtimes like crun and youki that support building with the Wasm runtime library. These two runtimes can also run regular Linux containers. Containerd uses `containerd-shim-runc-v2` to interface with low-level container runtimes.
```

Gauravpadam commented on content/en/blog/_posts/2024-01-26-webassembly-on-kubernetes-the-practice-guide/index.md:
````diff
@@ -0,0 +1,845 @@
+Run the following commands to install containerd:
+
+```shell
+export VERSION="1.7.3"
````
I recommend using the LTS release (which is v1.6 atm, I believe)

Gauravpadam commented on content/en/blog/_posts/2024-01-26-webassembly-on-kubernetes-the-practice-guide/index.md:
```diff
@@ -0,0 +1,845 @@
+
+### Run Wasm modules on Kubernetes
+
+To run Wasm workloads on Kubernetes, worker nodes need to be bootstrapped with a Wasm runtime, and RuntimeClass objects are used to assign workloads to nodes with Wasm support.
```
```suggestion
To run Wasm workloads on Kubernetes, worker nodes need to be bootstrapped with a Wasm runtime, and `RuntimeClass` objects are used to assign workloads to nodes with Wasm support.
```

sftim commented:
/retitle Add blog posts about WebAssembly in Cloud Native

sftim commented on content/en/blog/_posts/2024-01-26-webassembly-on-kubernetes-from-containers-to-wasm/index.md:
```diff
@@ -0,0 +1,73 @@
+
+An easy way to run Wasm modules in the container ecosystems is to embed the Wasm bytecode file in the Linux container image. Precisely, you can trim down the Linux OS inside the container to the point where it is just enough to support the Wasm runtime. Since Wasm modules are housed in standard containers, they can be integrated seamlessly with any existing container ecosystems.
+
+Compared to a regular Linux OS, the attack surface of the slimmed Linux OS is dramatically reduced. Nonetheless, this approach still necessitates the launching of a Linux container. Although the Linux OS is trimmed down, it still takes up 80% of the container's image size.
```
I'm worried about plagiarism. The text here seems close to wording in https://wasmedge.org/docs/develop/deploy/using-wasmedge-in-docker/

Can you explain the similarity?

cr7258 replied:
Here, I refer to the documentation of WasmEdge.
I rephrase the sentences: https://github.com/kubernetes/website/pull/43760/commits/f9582e75d9a7bf208c4c6ba408906a3fbcd57223

Gauravpadam commented on content/en/blog/_posts/2024-01-26-webassembly-on-kubernetes-the-practice-guide/index.md:
```diff
@@ -0,0 +1,845 @@
+After entering a shell on the node, follow the instructions in the section: [Run Wasm modules via low-level container runtimes](#run-wasm-modules-via-low-level-container-runtimes) to install the crun binary with Wasm support on the node.
+{{< /note >}}
+
+Modify the containerd configuration file `/etc/containerd/config.toml`, add the following content at the end:
+- Configure `crun` as the runtime handler for containerd. The format is `[plugins."io.containerd.grpc.v1.cri".containerd.runtimes.${HANDLER_NAME}]`.
+- `pod_annotations` allows passing the annotation `module.wasm.image/variant` to crun, which is set in Pod metadata to identify the Wasm workload. 
```
I suggest abbreviating this in a single toml code block

sftim commented on content/en/blog/_posts/2024-01-26-webassembly-on-kubernetes-the-practice-guide/index.md:
````diff
@@ -0,0 +1,845 @@
+```
+
+You can run Wasm modules through containerd:
+- `--runc-binary`: Specifies to use `crun` as low-level runtime to start the container.
+- `--runtime`: Specifies the version and name of the shim, which are converted by containerd into the binary name of the shim. For example: `io.containerd.runc.v2` → `containerd-shim-runc-v2`. Containerd starts the shim by running the `containerd-shim-runc-v2` binary file, which subsequently invokes `crun` to launch the container.
+- `--label`: Adds `"module.wasm.image/variant": "compat"` to indicate that this is a Wasm application without a guest OS.
````
I'm worried about plagiarism. This seems similar to https://bitone.dev/practical-guide-to-webAssembly-in-cloud-native/

Can you explain the similarity?

sftim reviewed (changes requested):
Let's address my concerns in:

- https://github.com/kubernetes/website/pull/43760#discussion_r1472938764
- https://github.com/kubernetes/website/pull/43760#discussion_r1472933431

Gauravpadam reviewed (commented):
Thanks @cr7258

This is obviously well-written, but I have some feedback

- The read time is still a good 15-20 minutes, any way we can bring this down with minimal content loss?
- I'd suggest refactoring places where code style is needed. I've marked some instances in my feedback

This needs a new publication date as the old one has passed.

sftim commented:
With few and **very specific** exceptions, we cannot accept articles that have been published elsewhere, even when the original is in a different language.

https://github.com/kubernetes/website/pull/43760#discussion_r1472967272 implies that this has been published elsewhere, and I believe you.

On that basis, we can't accept these articles. There is no relevant ground for an exception. However, if you publish a future article on this blog _first_ and then take a translation, or the same text, for publication elsewhere, that is fine (with an acknowledgement).

I am sorry about the frustration this must be for you @cr7258.

/close

0xE282B0 commented:
@cr7258, I'd still love to see this article published in english!
Could you copy it over to something like dev.to?

cr7258 commented:
@sftim It's a pity, but I can understand.  I really appreciate everyone who participated in reviewing this article.
The initial intention of submitting this article to the k8s blog was that after I published the Chinese version of this article, it was well received by readers.  Someone encouraged me to submit it to the k8s blog so that more people could read this article.

@0xE282B0 Sure, I'll post this article in English somewhere. I'll let you know.

sftim commented:
If you publish elsewhere, please ack the Kubernetes project; contributions toward the draft article are covered by our CLA. For example: “This article incorporates contributions and feedback from the Kubernetes project, which are copyright © 2024 The Linux Foundation”

cr7258 commented:
Hi @0xE282B0, I post the article on the CNCF blogs: https://www.cncf.io/blog/2024/03/12/webassembly-on-kubernetes-from-containers-to-wasm-part-01/
The part 02 (practice guide) will be published on March 26th.
